js/Update.elm

```diff
@@ -543,202 +541,153 @@ changeStage stage model =
 -- HELPER UPDATERS
 
 
-updateIfWelcome :
-    (WelcomeModel -> ( AppModel, Cmd Msg ))
-    -> AppModel
-    -> ( AppModel, Cmd Msg )
-updateIfWelcome upd model =
-    case model of
-        WelcomeScreen m ->
-            upd m
-
-        _ ->
-            (Debug.log
-                ("Tried running update function "
-                    ++ toString upd
-                    ++ " during "
-                    ++ toString model
-                )
-            )
-                ( model, Cmd.none )
+type alias Lens a b s t =
+    { get : s -> Maybe a
+    , set : b -> s -> t
+    }
```

Comment:
For example, here, instead of using `a b s t`, could you give them explicit names? When I'm making a generic type, I think it's okay to use one or maybe two one-letter type parameters, but four is pretty hard to remember.

E.g. can you use rename s -> key? a -> value? Should a and b be the same type? What is t, success/failure?

Reply:
I could, though it's abstract enough that I'm not sure what to put. Maybe I can explain the concept to you first, and we can figure out a naming.

Reply:
`Lens a b s t` means a way to zoom in on a value of type `s` that optionally contains a value of type `a`. For example, `s=GameModel` might contain a `a=TradeModel`, depending on what stage the game is at. This information is stored in the `get` field of the `Lens`.

Reply:
The `b` and `t` type variables represents the resulting types if we want to update `a`. If we run an update function `a->b` (for example, `updateAuction msg : AuctionModel -> (AuctionModel, Cmd Msg)`, then we should be able to induce an update `GameModel -> (GameModel, Cmd Msg)` on the parent model. This is achieved by specifying a setter `(AuctionModel, Cmd Msg) -> GameModel -> (GameModel, Cmd Msg)`.

Reply:
The only `b` and `t` we use in our code is when `b=Eff a=(a, Cmd Msg)` and `t = Eff s = (s, Cmd Msg)`, but I wrote it down in its full abstract form because it'll come in handy if one day we don't want to always return the top-level `Msg` type. And also because writing it this way makes implementing the functions easier for me.

Reply:
I think I understand why functional programming is slow to become popular, haha. I find it very difficult to read.

My instinct is still to give the generic fields names, e.g. `t-> SubmodelUpdate` and `b -> ModelUpdate`. That way the function signature might look like

```
get : Key -> Maybe Submodel
set : SubmodelUpdate -> Key -> ModelUpdate
```

If you think that the original version is more idiomatic, I guess, go for it. I would just say that it looks impenetrable to me!

Reply:
I think it's more important you can read it than to make it idiomatic. Maybe like this?
```
get : model -> Maybe submodel
set : updatedSubmodel -> model -> updatedModel
```

Reply:
That makes way more sense to me. Thanks for clarifying!